core/search/disjunction.hpp

```diff
@@ -614,6 +891,13 @@ doc_iterator::ptr make_disjunction(
       // empty or unreachable search criteria
       return doc_iterator::empty();
     case 1:
+      if /*constexpr*/ (std::is_same<typename Disjunction::doc_iterators_t::value_type,
+          position_score_iterator_adapter<typename Disjunction::doc_iterators_t::value_type::doc_iterator_t>>::value) {
+        typedef typename Disjunction::unary_disjunction_t unary_disjunction_t;
+        return doc_iterator::make<unary_disjunction_t>(
+          std::move(itrs.front())
+        );
+      }
```

Comment:
let's avoid using this logic based on adapter type, better to extend template so that one can control this logic